iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/project/tara/CachedTaraReferenceExtractionJob.java

```diff
@@ -0,0 +1,159 @@
+                    StructField$.MODULE$.apply("title", DataTypes.StringType, true, Metadata.empty()),
+                    StructField$.MODULE$.apply("abstract", DataTypes.FloatType, true, Metadata.empty()),
+                    StructField$.MODULE$.apply("text", DataTypes.StringType, true, Metadata.empty())
+            )
+    );
+
+    public static final StructType DOCUMENT_METADATA_WITH_HASH_SCHEMA = StructType$.MODULE$.apply(
+            Arrays.asList(
+                    StructField$.MODULE$.apply("id", DataTypes.StringType, false, Metadata.empty()),
+                    StructField$.MODULE$.apply("title", DataTypes.StringType, true, Metadata.empty()),
+                    StructField$.MODULE$.apply("abstract", DataTypes.FloatType, true, Metadata.empty()),
+                    StructField$.MODULE$.apply("text", DataTypes.StringType, true, Metadata.empty()),
+                    StructField$.MODULE$.apply("hashValue", DataTypes.IntegerType, false, Metadata.empty())
+            )
+    );
+
+    public static void main(String[] args) throws Exception {
+        JobParameters params = new JobParameters();
+        JCommander jcommander = new JCommander(params);
+        jcommander.parse(args);
+
+        logger.info("inputExtractedDocumentMetadata: {}", params.inputExtractedDocumentMetadataMergedWithOriginal);
+        logger.info("inputDocumentText: {}", params.inputDocumentText);
+        logger.info("lockManagerFactoryClassName: {}", params.lockManagerFactoryClassName);
+        logger.info("cacheRootDir: {}", params.cacheRootDir);
+        logger.info("scriptsDir: {}", params.scriptsDir);
+        logger.info("projectDbFile: {}", params.projectDbFile);
+        logger.info("numberOfEmittedFiles: {}", params.numberOfEmittedFiles);
+        logger.info("outputDocumentToProject: {}", params.outputDocumentToProject);
+
+        SparkConf conf = new SparkConf();
+
+        CacheMetadataManagingProcess cacheManager = new CacheMetadataManagingProcess();
+
+        try (SparkSession spark = SparkSessionFactory.withConfAndKryo(conf)) {
+            clearOutput(spark, params.outputDocumentToProject);
+
+            AvroDataFrameSupport avroDataFrameSupport = new AvroDataFrameSupport(spark);
+
+            Dataset<Row> inputExtractedDocumentMetadataMergedWithOriginalDF = avroDataFrameSupport.read(
+                    params.inputExtractedDocumentMetadataMergedWithOriginal,
+                    ExtractedDocumentMetadataMergedWithOriginal.SCHEMA$);
+            Dataset<Row> inputDocumentTextDF = avroDataFrameSupport.read(params.inputDocumentText,
+                    DocumentText.SCHEMA$);
+            Dataset<Row> documentMetadataDF = buildDocumentMetadata(inputDocumentTextDF,
+                    inputExtractedDocumentMetadataMergedWithOriginalDF);
+            Dataset<Row> documentMetadataWithHashDF = buildDocumentMetadataWithHash(documentMetadataDF)
+                    .cache();
```

Comment:
It's good that you've commented on using spark caching mechanism in this job because I've almost completely forgot to test it. 

For example I think we do not need to cache the result of hashing, so the line that this comment is attached to. The reason is hashing is implicitly persisted during shuffling. That is after the stage that does the hashing there is a shuffle which persists the results locally for following stages to fetch. The actual stage that reuses the hashes reads the files and does not recalculate the hashes. It can be seen in the spark ui as stages marked as skipped in this test run of the [job.](http://iis-cdh5-test-gw.ocean.icm.edu.pl:18089/history/application_1585920557248_45684/1/jobs/job/?id=2).

Other places where `.cache()` is used are also not well thought. Actually I think that cache should only be used for caching the results of running madis scripts. This causes the last action in this job (saving the results to output location) reuse cached results and not run pipe again. Also because the resulting dataframe is small (we have less then 300 results for TARA project mining) we can use cache with default storage level. This should cache the dataframe in memory.

I also think that your search query was not precise enough because there are only 3 places where `.cache()` is used :) Your search query must've match also with other places the string 'cache' is used like `params.cacheRootDir`.

Summarizing, there is now only one dataframe that will be cached - the results of reference extraction. It is cached with default storage level because it should fit in memory.

Reply:
OK, default storage level for dataframe is `MEMORY+DISK` which is more than OK. As you said the results should easily fit in memory but on this level we are 100% sure the madis won't be triggered twice because we will have the data also cached on disk (just in case due to some strange reason results won't fit in memory).